### PR TITLE
Optimize request templating for service requests

### PR DESCRIPTION
**Pull request by hxjuneja**

### CODE REVIEW

**Comment by hardikj** [comment]
Please hold a on this, this need to be updated.

File: lib/proxyHandler.js [hunk 1]
```diff
@@ -80,10 +80,49 @@ handler.prototype.parse_req = function(req, parentReq, res) {
...
+function _assemble(spec) {
+    var traverse = function(path) {
+        // A simple dotted path reference
+        var context = "req"
+        path.split(/\./).forEach(function(pathBit, index) {
+            if (index) {
+                context += "['"+pathBit+"']";
+            }
+        });
+        return "newReq['"+key+"'] = "+context+";";
```
Comment (gwicke) [inline lib/proxyHandler.js]: key needs escaping (use JSON.stringify)

File: lib/proxyHandler.js [hunk 2]
```diff
@@ -80,10 +80,49 @@ handler.prototype.parse_req = function(req, parentReq, res) {
...
-handler.prototype.eval_request = function(spec) {
-    return function (req) {
-        return this.parse_req(spec, req);
+function _assemble(spec) {
+    var traverse = function(path) {
+        // A simple dotted path reference
+        var context = "req"
+        path.split(/\./).forEach(function(pathBit, index) {
+            if (index) {
+                context += "['"+pathBit+"']";
```
Comment (gwicke) [inline lib/proxyHandler.js]: pathBit needs escaping

File: lib/proxyHandler.js [hunk 3]
```diff
@@ -80,10 +80,49 @@ handler.prototype.parse_req = function(req, parentReq, res) {
...
 };
 
-handler.prototype.eval_request = function(spec) {
-    return function (req) {
-        return this.parse_req(spec, req);
+function _assemble(spec) {
+    var traverse = function(path) {
+        // A simple dotted path reference
+        var context = "req"
+        path.split(/\./).forEach(function(pathBit, index) {
```
Comment (gwicke) [inline lib/proxyHandler.js]: We could actually require that identifiers conform to `[a-zA-Z_]+`. That would let us simply validate the path, and skip the rewriting.

File: lib/proxyHandler.js [hunk 4]
```diff
@@ -80,10 +80,49 @@ handler.prototype.parse_req = function(req, parentReq, res) {
...
+        });
+        return "newReq['"+key+"'] = "+context+";";
     };
+    var code = [];
+    if (spec === "$request") {
+        code.push("return req");
+    } else {
+        code.push("var newReq = {};");
+        for (var key in spec) {
+            if (/^\$/.test(spec[key])) {
```
Comment (gwicke) [inline lib/proxyHandler.js]: This could use some comments on which case this is.

File: lib/proxyHandler.js [hunk 5]
```diff
@@ -80,10 +80,49 @@ handler.prototype.parse_req = function(req, parentReq, res) {
     }
 };
 
-handler.prototype.eval_request = function(spec) {
-    return function (req) {
-        return this.parse_req(spec, req);
+function _assemble(spec) {
+    var traverse = function(path) {
+        // A simple dotted path reference
+        var context = "req"
```
Comment (gwicke) [inline lib/proxyHandler.js]: This is really code & not context, isn't it?

File: lib/router.js [hunk 6]
```diff
@@ -115,16 +116,18 @@ Router.prototype._loadModule = function (modDef) {
 
 function makeRequestTemplate(spec) {
     var uriTemplate = new URI(spec.uri, {}, true);
+    var reqEvaluator = new request_parser().eval_request
+    reqEvaluator = reqEvaluator(spec);
```
Comment (gwicke) [inline lib/router.js]: This is pretty confusing. Why is reqEvaluator re-assigned?

File: lib/proxyHandler.js (hunk 2 above)
Comment (hardikj) [inline lib/proxyHandler.js]: using encodeURIComponent right?

**Comment by hardikj** [comment]
closing this for https://github.com/wikimedia/restbase/pull/170/

File: lib/proxyHandler.js (hunk 2 above)
Comment (gwicke) [inline lib/proxyHandler.js]: JSON.stringify

**Pull request closed without merging**